File: 322-coin-change/coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    coins.sort((a, b) => b - a)

    const dp = Array(amount+1).fill(Infinity)
    dp[0] = 0

    for (let coin of coins) {
        for (let i = coin; i <= amount; i++) {
            dp[i] = Math.min(dp[i], dp[i - coin] + 1)
        }
    }

    return dp[amount] === Infinity ? -1 : dp[amount]
};

File: 322-coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], target: int) -> int:
        dp = [float('inf')] * (target + 1)
        dp[0] = 0

        for t in range(1, target + 1):
            for coin in coins:
                if coin <= t:
                    dp[t] = min(dp[t], 1 + dp[t-coin])
        
        return dp[target] if dp[target] != float('inf') else -1